Adding support for external C functions that have integer (or empty) args and/or returns

emarteca opened the pull request:
Starts addressing @https://github.com/rust-lang/miri/issues/11

### Implementation
Adding support for calling external C functions that have any number of integer arguments (types of integers: `i8`, `i16`, `i32`, `i64`, `u8`, `u16`, `u32`, `u64`) and an integer return type (or `void`).
As suggested in @https://github.com/rust-lang/miri/issues/11, the [`libffi` crate](https://docs.rs/libffi/latest/libffi/index.html) is used to dispatch the calls to external C functions.

#### Modifications
Main modifications are to:
* [helper.rs](src/helpers.rs) : adding a function `call_and_add_external_c_fct_to_context` to read the code pointer to the external C function, dispatch the call, and save the return in MIRI internal memory. Handles all conversions between MIRI and C values.
* [foreign_items.rs](src/shims/foreign_items.rs) : handles the calling of `call_and_add_external_c_fct_to_context` in [helper.rs](src/helpers.rs) when a foreign item is encountered. Also adds some structs to model C representations of arguments, and the signature of the external C call.

### Testing
Adds tests for the following external functions which are now supported:
* [int tests](tests/pass/external_C/int_c_tests.rs): 
     - adds 2 to a provided int (no type of int specified, so autocasts)
     - takes the sum of its 12 arguments (tests stack spill)
     - adds 3 to a 16 bit int
     - adds an `i16` to an `i64`
     - returns -10 as an unsigned int
* [void tests](tests/pass/external_C/print_from_c.rs)
     - void function that prints from C

saethlin committed:
Store protectors outside Item, pack Tag and Perm

Previously, Item was a struct of a NonZeroU64, an Option which was
usually unset or irrelevant, and a 4-variant enum. So collectively, the
size of an Item was 24 bytes, but only 8 bytes were used for the most
part.

So this takes advantage of the fact that it is probably impossible to
exhaust the total space of SbTags, and steals 3 bits from it to pack the
whole struct into a single u64. This bit-packing means that we reduce
peak memory usage when Miri goes memory-bound by ~3x. We also get CPU
performance improvements of varying size, because not only are we simply
accessing less memory, we can now compare a Vec<Item> using a memcmp
because it does not have any padding.

RalfJung committed:
adjust tests for removal of unsized_locals

RalfJung committed:
test for better error location on stack pop

RalfJung committed:
adjust code for copy_op changes

RalfJung committed:
move checking ptr tracking on item pop into cold helper function

oli-obk committed:
Implement strict comment parsing for ui tests

oli-obk committed:
Replace all //error-pattern comments with strict syntax and fix other comments in the same files

oli-obk committed:
Revert testing ui_test before ui tests

oli-obk committed:
Introduce a proper error handling framework

oli-obk committed:
Actually use eyre and get rid of the ad-hoc macros emulating error handling

oli-obk committed:
Document all the things

oli-obk committed:
Implement a proper command parser...

... that grabs things from the front instead of splitting at spaces and colons and hoping for the best

oli-obk committed:
rustfmt

RalfJung committed:
rustup

bors committed:
Auto merge of #2339 - RalfJung:rustup, r=RalfJung

rustup

RalfJung committed:
rustup; ptr atomics

bors committed:
Auto merge of #2341 - RalfJung:rustup, r=RalfJung

rustup; ptr atomics

Adds support for the operations added in https://github.com/rust-lang/rust/pull/96935.
I made the pointer-binops always return the provenance of the *left* argument; `@thomcc` I hope that is what you intended. I have honestly no idea if it has anything to do with what LLVM does...

I also simplified our pointer comparison code while I was at it -- now that *all* comparison operators support wide pointers, we can unify those branches.

RalfJung committed:
rustup

bors committed:
Auto merge of #2314 - RalfJung:unsized, r=RalfJung

adjust for removed unsized_locals

The Miri side of https://github.com/rust-lang/rust/pull/98831

oli-obk committed:
Remove an outdated comment

oli-obk committed:
Add some comments

oli-obk committed:
Explain `tests::init` function

oli-obk committed:
Create a proper normalization regex parser

oli-obk committed:
Rustfmt

RalfJung committed:
remove a dead optimization

bors committed:
Auto merge of #2342 - RalfJung:dead-opt, r=RalfJung

remove a dead optimization

This became dead code when I removed "Untagged".

RalfJung committed:
fix comment in ./miri

bors committed:
Auto merge of #2343 - RalfJung:comment, r=RalfJung

fix comment in ./miri

pointed out in https://github.com/rust-lang/miri/pull/2288#discussion_r915445912

RalfJung committed:
rustup

bors committed:
Auto merge of #2345 - RalfJung:rustup, r=RalfJung

rustup

oli-obk commented:
🤯 

did you mean to open this here? This is not the main miri repository

oli-obk committed:
Add FIXME

oli-obk committed:
Documentation

oli-obk committed:
Hide regular backtrace information from user-facing errors

oli-obk committed:
Port all tests

bors committed:
Auto merge of #2332 - rust-lang:slash_slash_at, r=oli-obk

More robust comment parsing

fixes #2170

I haven't ported the entire test suite yet. Once we've done that, I will remove the old parsing system (or in fact, turn them into errors so that accidental usage of old-style comments will be detected)

maurer commented:
> exploding_head
> 
> did you mean to open this here? This is not the main miri repository

We opened this over here for early code review - we'll be sharing a design doc for rust-lang#11 on the issue early next week covering the more complex part of this (syncing with pointers) for discussion.

This PR goes through the steps to support moving simple non-pointer types across FFI.

maurer commented on miri:
```diff
@@ -174,6 +174,10 @@ test|test-debug|bless|bless-debug)
     find_sysroot
+    # rebuild the C test shared object file if it's not there
+    if [ ! -f "tests/external_C/libtestlib.so" ]; then
+        gcc -shared -o tests/external_C/libtestlib.so tests/external_C/test.c
```
At a minimum this should be $CC, and we should consider using `cc` in a `build.rs` if appropriate.

emarteca replied:
good idea -- i've made a `CC` variable for now

maurer commented on src/bin/miri.rs:
```diff
@@ -487,6 +487,13 @@ fn main() {
+            if std::path::Path::new(&filename).exists() {
+                miri_config.external_c_so_file = Some(filename);
+            } else {
+                panic!("-Zmiri-external_c_so_file requires a valid path to a file");
```
Maybe "path {} does not exist"? The error right now sounds like an argument wasn't provided, which isn't the issue.

emarteca replied:
good point

maurer commented on src/bin/miri.rs:
```diff
@@ -487,6 +487,13 @@ fn main() {
+        } else if let Some(param) = arg.strip_prefix("-Zmiri-external_c_so_file=") {
+            let filename = param.to_string();
+            if std::path::Path::new(&filename).exists() {
+                miri_config.external_c_so_file = Some(filename);
```
Since we currently only allow one to be set (this is fine for now, we can add more later) maybe we should error out if `external_c_so_file` is already set to avoid confusion?

emarteca replied:
good idea

maurer commented on src/eval.rs:
```diff
@@ -126,6 +126,9 @@ pub struct MiriConfig {
     pub retag_fields: bool,
+    /// the location of a shared object file to load when calling external C functions
+    /// TODO! this is probably a temp solution
+    pub external_c_so_file: Option<String>,
```
Nit: `Option<PathBuf>`?

emarteca replied:
nice

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        for (cur_arg, arg_type) in args.iter().zip(external_fct_defn.inputs_types.iter()) {
+            match this.read_scalar(cur_arg) {
+                Ok(k) => {
+                    // the ints
```
This can be made more concise/maintainable by using macros. lmk if you want me to show you.

emarteca replied:
macros are a *major* improvement

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        }
+
+        // convert them to libffi Arg type
+        let libffi_args = libffi_args.iter()
```
We don't need to push a bunch of arguments onto a vector, just to iterate over them to produce a new vector. Consider:
```
for (cur_arg, arg_type) in args.iter().zip(external_fct_defn.inputs_types.iter()) {
  let (arg, c_arg) = extract_scalar(this.read_scalar(cur_arg), arg_type);
  libffi_args.push(arg_downcast(arg, c_arg));
}
```

emarteca replied:
good point -- note, as we talked about IRL we need to keep 2 loops because the libffi `arg` function takes a reference

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        use std::ops::Deref;
+
+        // get the function arguments, and convert them to libffi compatible form
+        let mut libffi_args = Vec::<(Box<dyn Any>, CArg)>::with_capacity(args.len());
```
Why are we using `Any` at all given that `CArg` can represent all the types you put into `Any`?

emarteca replied:
I swear there was a reason at some point -- but you're right, it's not needed :P

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        let libffi_args = libffi_args.iter()
+            .map(|cur_arg| {
+                match cur_arg {
+                    (boxed, CArg::Int8(_)) => {
```
Probably can be factored out into a method on `CArg` that projects into a libffi arg?

emarteca replied:
yes, good idea!

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+
+        // call the C function using libffi
+        unsafe {
+            let lib = match libloading::Library::new(match this.machine.external_c_so_file.as_ref(){
```
Nit: `.unwrap_or_else(|| throw_unsup_format!("..."))`

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+            let lib = match libloading::Library::new(match this.machine.external_c_so_file.as_ref(){
+                Some(x) => x,
+                None => {
+                    throw_unsup_format!("Error reading specified shared object file");
```
This should be "No shared object specified"

emarteca replied:
good catch

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    throw_unsup_format!("Error reading specified shared object file");
+                }
+            }) {
+                Ok(x) => x,
```
`unwrap_or_else`

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                }
+            };
+            let func: libloading::Symbol<'_, unsafe extern fn()> = match lib.get(link_name.as_str().as_bytes()) {
+                Ok(x) => x,
```
`unwrap_or_else`

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+
+        // call the C function using libffi
+        unsafe {
+            let lib = match libloading::Library::new(match this.machine.external_c_so_file.as_ref(){
```
Doesn't need to be this change, but the loading of the library and function resolution should be cached.

emarteca replied:
good point: this is fixed by storing the library in the `machine` instead of the path to the library (as you suggested)

maurer commented on src/helpers.rs:
```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+            // code pointer to C function
+            let ptr = CodePtr(*func.deref() as *mut _);
+            // store output, depending on return type in the function signature
+            match external_fct_defn.output_type {
```
Factor out into a helper, consider macros to reduce duplication.

emarteca replied:
good point

maurer commented on src/machine.rs:
```diff
@@ -339,6 +339,9 @@ pub struct Evaluator<'mir, 'tcx> {
     pub(crate) since_progress_report: u32,
+
+    /// Path to the optional C shared object file
+    pub external_c_so_file: Option<String>,
```
Maybe `Option<libloading::Library>` to avoid reloads?

emarteca replied:
good idea!

maurer commented on src/shims/foreign_items.rs:
```diff
@@ -351,13 +424,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        if let Some(local_id) = def_id.as_local() {
+            if let Some(extern_c_fct_rep) = ExternalCFuncDeclRep::from_hir_node(
+                &tcx.hir().get(tcx.hir().local_def_id_to_hir_id(local_id)), link_name) {
+                    if let Ok(_) = this.call_and_add_external_c_fct_to_context( extern_c_fct_rep, dest, args) {
```
There's an extra space here - did you run `rustfmt`?

maurer commented on src/shims/foreign_items.rs:
```diff
@@ -351,13 +424,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         ret: mir::BasicBlock,
     ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
         let this = self.eval_context_mut();
+        
```
Consider checking this after we check for shims?

Probably by default we want summaries if present, then the real world otherwise.

emarteca replied:
I had thought we'd want any user-provided code to take precedence over the shims.

saethlin committed:
Pass through isatty if the host is also unix

maurer reviewed (commented):
Looks like a good start, the main thing I think we want to do is factoring several of the large functions to use helper functions/methods, since as is things are going to grow too large to reason about pretty quickly.

RalfJung committed:
ui_test: require colon after command

RalfJung committed:
rename ui_test::comments → parser

RalfJung committed:
typo

Co-authored-by: Oli Scherer <github35764891676564198441@oli-obk.de>

bors committed:
Auto merge of #2348 - RalfJung:ui-test, r=RalfJung

ui_test: require colon after command

Turns out we already use colons everywhere properly.

Also rename ui_test::comments to parser. I think that better describes its contents. :)

RalfJung committed:
remove deprecated options

RalfJung committed:
extend a comment in readlink

bors committed:
Auto merge of #2351 - RalfJung:readlink, r=RalfJung

extend a comment in readlink

saethlin committed:
Add a benchmark of the hang-on-test-failure code path

Techie-Pi committed:
Fix README typo

"behaivours" -> "behaviours".
Just something I noticed while reading

RalfJung committed:
make a test deterministic

RalfJung committed:
make more tests deterministic

bors committed:
Auto merge of #2355 - RalfJung:rustup, r=RalfJung

make a test deterministic

bors committed:
Auto merge of #2354 - Techie-Pi:patch-1, r=RalfJung

Fix README typo

"behaivours" -> "behaviours".
Just something I noticed while reading

oli-obk committed:
require level and colon in `//~` style comments

oli-obk committed:
Remove error annotations stderr

they still showed up in multiline messages

oli-obk committed:
Pacify clippy

oli-obk committed:
Fix annotation filtering with revisions

oli-obk committed:
Document require-annotations-for-level

bors committed:
Auto merge of #2356 - rust-lang:slash_slash_at, r=RalfJung

require level and colon in `//~` style comments

implements https://github.com/rust-lang/miri/pull/2348#issuecomment-1179551953

emarteca commented:
Thanks for the feedback! I've pushed commits addressing the comments. As you state, the main change is factoring out functionality into helper methods, and [introducing macros](https://github.com/maurer/miri/pull/1/files/399c85d7a609ef1fa6ffe4e8c6b39d6085dcb311..206de3f9295fcc33fadeb939e794193f065ce27a#diff-c2a0fff5575ebf83634ea4da334a24d52775bcef6700ed6f9d0bc4e3db79e42aR81) to cut down on the code duplication (hooray for macros!! 🎉 )

saethlin committed:
Rearrange and document the new implementation

stacked_borrow now has an item module, and its own FrameExtra. These
serve to protect the implementation of Item (which is a bunch of
bit-packing tricks) from the primary logic of Stacked Borrows, and the
FrameExtra we have separates Stacked Borrows more cleanly from the
interpreter itself.

The new strategy for checking protectors also makes some subtle
performance tradeoffs, so they are now documented in Stack::item_popped
because that function primarily benefits from them, and it also touches
every aspect of them.

Also separating the actual CallId that is protecting a Tag from the Tag
makes it inconvienent to reproduce exactly the same protector errors, so
this also takes the opportunity to use some slightly cleaner English in
those errors. We need to make some change, might as well make it good.

RalfJung committed:
set up triagebot

RalfJung committed:
allow some relabeling as well

bors committed:
Auto merge of #2359 - RalfJung:triagebot, r=RalfJung

set up triagebot

`@Mark-Simulacrum` does this look like a sensible configuration?

RalfJung committed:
test that we also find bad uses of mem::uninitialized

bors committed:
Auto merge of #2361 - RalfJung:uninit, r=RalfJung

test that we also find bad uses of mem::uninitialized

Also we really don't need to separately test signed and unsigned integers... our test suite is big enough as it is. ;)

maurer commented on miri:
```diff
@@ -56,6 +56,11 @@ EOF
 
+# if $CC is not set, use gcc as default
+if [[ -z $CC ]]; then
+  CC=gcc
```
Maybe use `cc` as default?

emarteca replied:
sure!

maurer reviewed (approved):
Looks reasonable to me.

bors committed:
Auto merge of #2315 - saethlin:shrink-item, r=saethlin

Optimizing Stacked Borrows (part 2): Shrink Item

This moves protectors out of `Item`, storing them both in a global `HashSet` which contains all currently-protected tags as well as a `Vec<SbTag>` on each `Frame` so that when we return from a function we know which tags to remove from the protected set.

This also bit-packs the 64-bit tag and the 2-bit permission together when they are stored in memory. This means we theoretically run out of tags sooner, but I doubt that limit will ever be hit.

Together these optimizations reduce the memory footprint of Miri when executing programs which stress Stacked Borrows by ~66%. For example, running a test with isolation off which only panics currently peaks at ~19 GB, with this PR it peaks at ~6.2 GB.

To-do
- [x] Enforce the 62-bit limit
- [x] Decide if there is a better order to pack the tag and permission in
- [x] Wait for `UnsafeCell` to become infectious, or express offsets + tags in the global protector set

Benchmarks before:
```
Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/backtraces/Cargo.toml
  Time (mean ± σ):      8.948 s ±  0.253 s    [User: 8.752 s, System: 0.158 s]
  Range (min … max):    8.619 s …  9.279 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/mse/Cargo.toml
  Time (mean ± σ):      2.129 s ±  0.037 s    [User: 1.849 s, System: 0.248 s]
  Range (min … max):    2.086 s …  2.176 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/serde1/Cargo.toml
  Time (mean ± σ):      3.334 s ±  0.017 s    [User: 3.211 s, System: 0.103 s]
  Range (min … max):    3.315 s …  3.352 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/serde2/Cargo.toml
  Time (mean ± σ):      3.316 s ±  0.038 s    [User: 3.207 s, System: 0.095 s]
  Range (min … max):    3.282 s …  3.375 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/unicode/Cargo.toml
  Time (mean ± σ):      6.391 s ±  0.323 s    [User: 5.928 s, System: 0.412 s]
  Range (min … max):    6.090 s …  6.917 s    5 runs
 ```
 After:
 ```
Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/backtraces/Cargo.toml
  Time (mean ± σ):      6.955 s ±  0.051 s    [User: 6.807 s, System: 0.132 s]
  Range (min … max):    6.900 s …  7.038 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/mse/Cargo.toml
  Time (mean ± σ):      1.784 s ±  0.012 s    [User: 1.627 s, System: 0.156 s]
  Range (min … max):    1.772 s …  1.797 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/serde1/Cargo.toml
  Time (mean ± σ):      2.505 s ±  0.095 s    [User: 2.311 s, System: 0.096 s]
  Range (min … max):    2.405 s …  2.603 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/serde2/Cargo.toml
  Time (mean ± σ):      2.449 s ±  0.031 s    [User: 2.306 s, System: 0.100 s]
  Range (min … max):    2.395 s …  2.467 s    5 runs

Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/unicode/Cargo.toml
  Time (mean ± σ):      3.667 s ±  0.110 s    [User: 3.498 s, System: 0.140 s]
  Range (min … max):    3.564 s …  3.814 s    5 runs
 ```
 The decrease in system time is probably due to spending less time in the page fault handler.

oli-obk committed:
Error patterns can be regexes

bors committed:
Auto merge of #2352 - saethlin:new-benchmark, r=RalfJung

Add a benchmark of the hang-on-test-failure code path

This is the code pattern that produces the performance problem in https://github.com/rust-lang/miri/issues/2273

I figured out what I was stuck on in https://github.com/rust-lang/miri/pull/2315#discussion_r916387919. For a while I was just doing `let x: &[u8] = &[0u8; 4096];` but that doesn't produce the runtime inside `Stack::item_popped` that I was looking for, I think because this allocation is never deallocated. But with `Vec`, I get the profile I'm looking for.

bors committed:
Auto merge of #2364 - rust-lang:regex_error_annotations, r=RalfJung

Error patterns can be regexes

fixes #2362

RalfJung committed:
use ui_test regex capabilities

bors committed:
Auto merge of #2367 - RalfJung:ui-test-regex, r=RalfJung

use ui_test regex capabilities

We can have more precise patterns now, so use them for some of the tests where that makes sense. :)

RalfJung committed:
don't forcefully enable debug assertions, but make -debug mode usable

still set those flags on CI though, we want to catch overflow there

RalfJung committed:
move stacked_borrows.rs together with the other files of its module

RalfJung committed:
reborrow error: clarify that we are reborrowing *from* that tag

bors committed:
Auto merge of #2369 - RalfJung:reborrow, r=RalfJung

reborrow error: clarify that we are reborrowing *from* that tag

`@saethlin` I found the current message not entirely clear, so what do you think about this?

RalfJung committed:
add another weak mem consistency test

bors committed:
Auto merge of #2328 - RalfJung:perf, r=RalfJung

move checking ptr tracking on item pop into cold helper function

Before:
```
Benchmark 1: cargo miri run --manifest-path bench-cargo-miri/serde1/Cargo.toml
  Time (mean ± σ):      6.729 s ±  0.050 s    [User: 6.608 s, System: 0.124 s]
  Range (min … max):    6.665 s …  6.799 s    5 runs

Benchmark 2: cargo miri run --manifest-path bench-cargo-miri/unicode/Cargo.toml
  Time (mean ± σ):     20.923 s ±  0.271 s    [User: 20.386 s, System: 0.537 s]
  Range (min … max):   20.580 s … 21.165 s    5 runs
```
After:
```
Benchmark 1: cargo miri run --manifest-path bench-cargo-miri/serde1/Cargo.toml
  Time (mean ± σ):      6.562 s ±  0.023 s    [User: 6.430 s, System: 0.135 s]
  Range (min … max):    6.544 s …  6.594 s    5 runs

Benchmark 2: cargo miri run --manifest-path bench-cargo-miri/unicode/Cargo.toml
  Time (mean ± σ):     20.375 s ±  0.228 s    [User: 19.964 s, System: 0.413 s]
  Range (min … max):   20.201 s … 20.736 s    5 runs
```
Nothing major, but we'll take it I guess. :shrug:

Fixes https://github.com/rust-lang/miri/issues/2132

RalfJung committed:
rustup

oli-obk committed:
Allow tests to have dependencies

RalfJung committed:
fix a missing thread join

bors committed:
Auto merge of #2371 - RalfJung:rustup, r=RalfJung

rustup

Cc https://github.com/rust-lang/rust/issues/99224

RalfJung committed:
make cache consistency checks into regular debug assertions

RalfJung committed:
add work-around for #2164

RalfJung committed:
replace a macro by a function

RalfJung committed:
test that we can see this weak behavior

bors committed:
Auto merge of #2370 - RalfJung:weak-mem-test, r=RalfJung

add another weak mem consistency test

`@cbeuw` your code seems to do this correctly? :D

RalfJung committed:
rustup

bors committed:
Auto merge of #2372 - RalfJung:rustup, r=RalfJung

rustup

Cc https://github.com/rust-lang/rust/issues/99224

RalfJung committed:
use the cargo default for debug/release builds

RalfJung committed:
make some debug assertions in RangeObjectMap be full assertions

RalfJung committed:
ensure that RangeMap panics on OOB

RalfJung committed:
clarify how to run the manually installed Miri

RalfJung committed:
rustup

bors committed:
Auto merge of #2375 - RalfJung:rustup, r=RalfJung

Rustup

bors committed:
Auto merge of #2368 - RalfJung:debug, r=oli-obk

Make "./miri {build,run,test}" use debug assertions but "./miri install" not

This makes `./miri run`/`./miri test` use the full set of debug assertions (including the rather expensive ones that check consistency of the Stacked Borrows cache), but `./miri install` installs a Miri *without* those debug assertions.

That's the same behavior as cargo, and helps catch Miri bugs with the test suite while making installed Miri usable for larger runs.

RalfJung committed:
rustup

RalfJung committed:
make unused flags work like they used to

oli-obk committed:
Remove `--offline` from `./miri install` as otherwise we can't add more dependencies to miri as CI will fail.

if you want this locally, pass it to your invocation.

oli-obk committed:
Add test flag for running a test only on the host

oli-obk committed:
clippy

bors committed:
Auto merge of #2377 - rust-lang:only-on-host, r=oli-obk

Add test flag for running a test only on the host

bors committed:
Auto merge of #2376 - RalfJung:rustup, r=RalfJung

rustup

RalfJung committed:
use PlaceTy visitor

RalfJung committed:
remove a fastpath that does not seem to actually help

bors committed:
Auto merge of #2378 - RalfJung:sb, r=RalfJung

use PlaceTy visitor and dedup sime retagging code

I benchmarked this and as far as I can see the difference to the old code is totally within noise. And this makes the code a lot simpler and removes duplication so yay. :)

saethlin committed:
Test that isatty doesn't crash

bors committed:
Auto merge of #2349 - saethlin:isatty, r=RalfJung

Improve isatty support

Per https://github.com/rust-lang/miri/issues/2292#issuecomment-1171858283, this is an attempt at

> do something more clever with Miri's `isatty` shim

Since Unix -> Unix is very simple, I'm starting with a patch that just does that. Happy to augment/rewrite this based on feedback.

The linked file in libtest specifically only supports stdout. If we're doing this to support terminal applications, I think it would be strange to support one but not all 3 of the standard streams.

The `atty` crate contains a bunch of extra logic that libtest does not contain, in order to support MSYS terminals: https://github.com/softprops/atty/commit/db8d55f88eabfcc22cde39533c2b1d688d1cdfc6 so I think if we're going to do Windows support, we should probably access all that logic somehow. I think it's pretty clear that the implementation is not going to change, so I think if we want to, pasting the contents of the `atty` crate into Miri is on the table, instead of taking a dependency.

RalfJung committed:
with isolation we want to be fully deterministic

RalfJung committed:
libc test: also call isatty on an actual file

RalfJung committed:
ui_test: more robust syntax for target filtering

RalfJung committed:
make an expect into proper error reporting

bors committed:
Auto merge of #2380 - RalfJung:isatty, r=RalfJung

isatty tweaks

`@saethlin` I realized we need to gate this behind the isolation flag, sorry. :/  The point of isolated mode is that it is 100% deterministic, so we cannot call host functions that might give different answers on different runs.

bors committed:
Auto merge of #2379 - RalfJung:ui-test-target-filter, r=oli-obk

ui_test: more robust syntax for target filtering

Implicit fallbacks are always fragile and prone to typos, so let's be explicit.

oli-obk committed:
Add a dedicated thread for output printing

oli-obk committed:
Print one character per test instead of one line

RalfJung committed:
move atomic intrinsics to their own file

RalfJung committed:
move simd intrinsics to their own file

RalfJung committed:
remove ret param from foreign_item hierarchy

RalfJung committed:
add support for new RMW orders

RalfJung committed:
make atomic intrinsic impl details private

RalfJung committed:
refactor atomic intrinsic handling to actually parse the intrinsic name

bors committed:
Auto merge of #2382 - RalfJung:atomic, r=RalfJung

add support for new RMW orders

Implements support for https://github.com/rust-lang/rust/pull/98383.

Also restructure intrinsic shims a bit so we have atomic and SIMD intrinsics in their own file, and refactor atomic intrinsic name parsing.

Commit:
C FFI support for functions with int args and returns (rebasing from current master)

Commit:
updates to foreign_items

RalfJung committed:
use env vars, not Cargo.toml, to configure out dev profile

bors committed:
Auto merge of #2385 - RalfJung:dev, r=RalfJung

use env vars, not Cargo.toml, to configure out dev profile

This should fix the cargo warnings we are seeing in rustc.

RalfJung committed:
cargo-miri: write the Xargo project files atomically

RalfJung committed:
update ui_test readme

bors committed:
Auto merge of #2387 - RalfJung:ui-readme, r=oli-obk

update ui_test readme

I forgot to do that when changing the ignore/only syntax.

bors committed:
Auto merge of #2386 - RalfJung:xargo-atomic, r=RalfJung

cargo-miri: write the Xargo project files atomically

`@matthiaskrgr` this should fix https://github.com/rust-lang/miri/issues/2340. Any chance you could test this before we land it?

RalfJung committed:
rustup

bors committed:
Auto merge of #2388 - RalfJung:rustup, r=RalfJung

rustup

Cc https://github.com/rust-lang/rust/issues/99461

bors committed:
Auto merge of #2350 - RalfJung:remove-deprecated, r=oli-obk

remove deprecated options

These have been deprecated a month ago and nobody said they need them. Is that enough time? We can also let this PR sit a little.

Cc https://github.com/rust-lang/miri/issues/2187 https://github.com/rust-lang/miri/issues/2188
(keeping them open to track removing their supporting infrastructure in the core interpreter)

RalfJung committed:
show a better error when running Miri with the wrong sysroot

RalfJung committed:
show where the interpreter was interpreting when an ICE occurs

bors committed:
Auto merge of #2390 - RalfJung:debug, r=RalfJung

fix some debugging papercuts

- detect when Miri got locally installed and is being run with the wrong toolchain
- when an ICE occurs, print a backtrace of where the interpreter was at the time

RalfJung committed:
on an error, always print the unnormalized stderr

RalfJung committed:
tweak failure output a little

RalfJung committed:
rustup

bors committed:
Auto merge of #2392 - RalfJung:rustup, r=RalfJung

rustup

saethlin committed:
Fix bugs where unique_range became invalid

And also expand the cache integrity checks to cover this case, and
generally assert a lot more about the unique_range, then tighten up
sloppy implementation scenarios that this uncovered.

RalfJung committed:
add some dyn upcasting tests

bors committed:
Auto merge of #2395 - RalfJung:upcast, r=RalfJung

add some dyn upcasting tests

Copied from the rustc test suite.

saethlin committed:
Add a minimal reproducer of the ICE

Co-authored-by: Ralf Jung <post@ralfj.de>

saethlin committed:
Clarify implementation a bit

bors committed:
Auto merge of #2394 - saethlin:unique-range-ice, r=RalfJung

Fix bugs where unique_range became invalid

And also expand the cache integrity checks to cover this case.

I'm going to run this over all the ICEs I've gotten out of Miri recently, could be a bit.

Fixes https://github.com/rust-lang/miri/issues/2389

RalfJung committed:
only do env var cleanup if all threads have stopped

bors committed:
Auto merge of #2397 - RalfJung:cleanup, r=RalfJung

only do env var cleanup if all threads have stopped

Hopefully fixes https://github.com/rust-lang/miri/issues/2396

sunshowers committed:
[cargo-miri] support nextest

Add the ability to run the `list` and `run` nextest commands, which
enable per-test isolation.

Commit:
PR edits requested by @RalfJung: main changes are removing the use of `hir` types, some refactoring/renaming of functions, and edits to the related test file hierarchy

RalfJung committed:
no need for an exhaustive enum of subcommands

bors committed:
Auto merge of #2399 - RalfJung:subcommands, r=RalfJung

no need for an exhaustive enum of subcommands

oli-obk committed:
Add a scheme for always using the default toolchain, running clippy and fmt before running any other command

bors committed:
Auto merge of #2391 - RalfJung:stderr, r=oli-obk

on an error, always print the unnormalized stderr

Currently we skip this if a stderr diff was printed, but the stderr diff is normalized, so e.g. one cannot learn line numbers from it.

Is there a way to get the diff to only print the parts the differ, like a usual `diff -u`? Currently it always seems to print the full output, so with a long stderr we now get doubly long test results even if the diff is actually rather small.

RalfJung committed:
fix test file name

bors committed:
Auto merge of #2401 - RalfJung:dyn, r=RalfJung

fix test file name

Would be better to run this test when doing `./miri test dyn`.

bors committed:
Auto merge of #2373 - rust-lang:test_dependencies, r=RalfJung

Allow ui tests to have dependencies in a reliable way

This completely sidesteps the issue that compiletest-rs has where old artifacts of a dependency cause `multiple available crates of name XXX` errors. At this point I think we've reached feature parity for clippy, too, so I'm going to try publishing a version once this is merged.

Commit:
cfg for linux, and remove test SO file

RalfJung committed:
cargo-miri: reorder --target to after the user-defined commands

bors committed:
Auto merge of #2402 - RalfJung:cargo-target, r=RalfJung

cargo-miri: reorder --target to after the user-defined commands

This should help with https://github.com/rust-lang/miri/pull/2398.

RalfJung committed:
rustup

bors committed:
Auto merge of #2403 - RalfJung:rustup, r=RalfJung

rustup

Commit:
dont need to track `fn_sig`: nice simplification, thanks @oli-obk!

RalfJung committed:
check for extern static size mismatches

RalfJung committed:
handle get_alloc_extra the same throughout Stacked Borrows

Commit:
update README; moving libloading load and hack into its own function

RalfJung committed:
some stray tag → prov renames

bors committed:
Auto merge of #2404 - RalfJung:mix, r=RalfJung

check for extern static size mismatches

Also handle get_alloc_extra the same throughout Stacked Borrows. I don't think this `unwrap` can actually cause an ICE since another part of SB will raise an error before, but still, seems strange to do this inconsistently in retagging vs expose_ptr.

RalfJung committed:
fix miri script

bors committed:
Auto merge of #2381 - rust-lang:infra, r=RalfJung

Add a scheme for always using the default toolchain, running clippy and fmt before running any other command

I keep forgetting to run rustup-toolchain on rebases across toolchain updates

I also keep forgetting to run rustfmt and clippy. The former isn't run by vscode if I don't explicitly save (I have autosave on).

bors committed:
Auto merge of #2398 - sunshowers:nextest-compat, r=RalfJung

[cargo-miri] support nextest

Add the ability to run `cargo miri nextest list` and `cargo miri nextest run`.

[cargo-nextest](https://nexte.st) is a new test runner for Rust maintained mostly by myself. It has several new features, but the most relevant to miri is the fact that it runs [each test in its own process](https://nexte.st/book/how-it-works.html#the-nextest-model). This gives miri users better leak detection (https://github.com/rust-lang/miri/issues/1481) for free, for example.

See https://github.com/nextest-rs/nextest/issues/181 for discussion, including comments by `@eddyb` and `@RalfJung.`

Future work might be to have miri read [the list of tests](https://docs.rs/nextest-metadata/latest/nextest_metadata/struct.TestListSummary.html) (or [test binaries](https://docs.rs/nextest-metadata/latest/nextest_metadata/struct.BinaryListSummary.html)) generated by `nextest list`. `@eddyb` thinks that might be useful.

I tested `cargo miri nextest run` against smallvec, and it worked great.

Note: Running tests out of archives is currently broken, as the comment in run-test.py explains.

oli-obk committed:
Use names suggestive of channel endpoints

bors committed:
Auto merge of #2383 - rust-lang:dot_mode, r=oli-obk

Print one . character per test instead of one line

`./miri bless -- --quiet` now prints a dot per test, along with the regular Rust unit tests that listen to this flag

RalfJung committed:
fix auto-toolchain pwd

bors committed:
Auto merge of #2405 - RalfJung:pwd, r=RalfJung

fix auto-toolchain pwd

`rustup-toolchain` needs to be called in the right directory

oli-obk committed:
Avoid rustformatting on autosave

RalfJung committed:
make test-cargo-miri only about cargo

remove rand (large dependency) and page-size (testing the dependency, not cargo-miri).
keep only byteorder as a "demo" dependency, it is a leaf and builds quickly.

RalfJung committed:
test page_size via test dependency support

and move crate tests to their own folders

oli-obk committed:
Use a MIRI namespaced env var name for auto ops

RalfJung committed:
cargo-miri: set RUSTC to us

RalfJung committed:
only complain about runtime toolchain mismatch when there actually is a runtime toolchain

RalfJung committed:
reduce chance of RUSTC collisions

RalfJung committed:
some more debug output

RalfJung committed:
make the find_miri returned path actually exist

bors committed:
Auto merge of #2409 - RalfJung:cargo-miri-rustc, r=RalfJung

cargo-miri: set RUSTC to us

Works around https://github.com/rust-lang/cargo/issues/10885.

RalfJung committed:
aarch inline asm is not stable yet

bors committed:
Auto merge of #2408 - RalfJung:deps, r=oli-obk

make test-cargo-miri only about cargo

Move the things that actually test dependency behavior into the regular test suite, now that we can do that. :)

oli-obk committed:
Document MIRI_AUTO_OPS

bors committed:
Auto merge of #2410 - rust-lang:miri_auto_ops, r=RalfJung

Some cleanups and docs around the auto ops

Commit:
fix outdated doc

Commit:
upstream master changes; and no libc on windows

Commit:
upstream

Commit:
upstream

Commit:
upstream

Commit:
check status of cc for SO file

maurer commented:
Closing in favor of the real PR
